Sources/Valet/MigratableKeyValuePair.swift

```diff
@@ -0,0 +1,133 @@
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+//
+
+import Foundation
+
+/// A struct that represented a key:value pair that can be migrated.
+public struct MigratableKeyValuePair<KeyType: Hashable>: Hashable {
+
+    // MARK: Initialization
+
+    /// Creates a migratable key:value pair with the provided inputs.
+    /// - Parameters:
+    ///   - key: The key in the key:value pair.
+    ///   - value: The value in the key:value pair.
+    public init(key: KeyType, value: Data) {
+        self.key = key
+        self.value = value
+    }
+
+    /// Creates a migratable key:value pair with the provided inputs.
+    /// - Parameters:
+    ///   - key: The key in the key:value pair.
+    ///   - value: The desired value in the key:value pair, represented as a String.
+    public init(key: KeyType, value: String) {
+        self.key = key
+        self.value = Data(value.utf8)
+    }
+
+    // MARK: Public
+
+    /// The key in the key:value pair.
+    public let key: KeyType
+    /// The value in the key:value pair.
+    public let value: Data
+}
+
+// MARK: - Objective-C Compatibility
+
+@objc(VALMigratableKeyValuePairInput)
+public final class ObjectiveCCompatibilityMigratableKeyValuePairInput: NSObject {
+
+    // MARK: Initialization
+
+    internal init(key: Any, value: Data) {
+        self.key = key
+        self.value = value
+    }
+
+    // MARK: Public
+
+    /// The key in the key:value pair.
+    @objc
+    public let key: Any
+    /// The value in the key:value pair.
+    @objc
+    public let value: Data
+}
+
+@objc(VALMigratableKeyValuePairOutput)
+public class ObjectiveCCompatibilityMigratableKeyValuePairOutput: NSObject {
+
+    // MARK: Initialization
+
+    /// Creates a migratable key:value pair with the provided inputs.
+    /// - Parameters:
+    ///   - key: The key in the key:value pair.
+    ///   - value: The value in the key:value pair.
+    @objc
+    public init(key: String, value: Data) {
+        self.key = key
+        self.value = value
+        preventMigration = false
+    }
+
+    /// Creates a migratable key:value pair with the provided inputs.
+    /// - Parameters:
+    ///   - key: The key in the key:value pair.
+    ///   - stringValue: The desired value in the key:value pair, represented as a String.
+    @objc
+    public init(key: String, stringValue: String) {
+        self.key = key
+        self.value = Data(stringValue.utf8)
+        preventMigration = false
+    }
+
+    // MARK: Public Static Methods
+
+    /// A sentinal `ObjectiveCCompatibilityMigratableKeyValuePairOutput` that conveys that the migration should be prevented.
+    @objc
+    public static func preventMigration() -> ObjectiveCCompatibilityMigratableKeyValuePairOutput {
```

Comment:
This method sounds to me more than its actually doing the action to prevent migration on an instance rather than creating an instance that prevents the migration.

Maybe `migrationDisabled` / `migrationPrevented`?

Reply:
Yeah the tough thing here is it isn't super clear that we're doing this work _before_ the migration has started. This is why I wanted to use a future tense. A past-tense phrase here might make it read as if returning it prevented/disabled the migration of just this pair (or future pairs).

In the context where it's used, I hope this method reads well:
https://github.com/square/Valet/blob/c7cec7516d89d82768f5cff14b2d599d54304871/Tests/ValetObjectiveCBridgeTests/VALValetTests.m#L475-L484

The documentation on the `migrateObjectsMatching` Objective-C method also makes it clear how to use this method:
https://github.com/square/Valet/blob/c7cec7516d89d82768f5cff14b2d599d54304871/Sources/Valet/Valet.swift#L661-L668

While I agree that reading this method by itself is a bit confusing, I think within the context it's supposed to be used it makes sense? Let me know if not @fdiaz.

Reply:
That said, I should make this method available only to Objective-C